Added initial design doc for cross platform (#5628)
dotnet/machinelearning: docs/code/CrossPlatform.md

		- [3.4 Hybrid Between Finding Replacements and Software Fallbacks](#34-hybrid-between-finding-replacements-and-software-fallbacks)
	- [My Suggestion](#my-suggestion)
		- [NET Core 5](#net-core-5)
		- [3rd Party Dependencies](#3rd-party-dependencies)
		- [Helix](#helix)
		- [Mobile Support](#mobile-support)
		- [Support Grid](#support-grid)

## 1. Why cross-platform/architecture
ML.NET is an open-source machine learning framework which makes machine learning accessible to .NET developers with the same code that powers machine learning across many Microsoft products, including Power BI, Windows Defender, and Azure.

ML.NET allows .NET developers to develop/train their own models and infuse custom machine learning into their applications using .NET.

Currently, while .NET is able to run on many different platforms and architectures, ML.NET is only able to run on Windows, Mac, and Linux, either x86 or x64. This excludes many architectures such as Arm, Arm64, M1, and web assembly, places that .NET is currently able to run.

The goal is to enable ML.NET to run everywhere that .NET itself is able to run.

## 2. Current status
### 2.1 Problems

> ML.NET has a hard dependency on Intel MKL. It performs many optimized math functions and enables several transformers/trainers to be run natively for improved performance. The problem is that Intel MKL can only run on x86/x64 machines and is the main blocker for expanding to other architectures.

Reviewer: You did a good job at highlighting the hard problems.  Can we state the higher level problems as well, to frame things?
1. We only build and test for a subset of platforms that .NET supports. 
2. Native components must be explicitly built for additional architectures which we wish to support.  This limits our ability to support new platforms without doing work. 
3. Building our native components for new platforms faces challenges due to lack of support for those components dependencies.  This limits our ability to support the current set of platforms.
4. Some of our external dependencies have limited support for .NET's supported platforms.

> ML.NET has a hard dependency on Intel MKL. It performs many optimized math functions and enables several transformers/trainers to be run natively for improved performance. The problem is that Intel MKL can only run on x86/x64 machines and is the main blocker for expanding to other architectures.

Reviewer 2: ```suggestion
Parts of ML.NET have a hard dependency on Intel MKL. It performs many optimized math functions and enables several transformers/trainers to be run natively for improved performance. The problem is that Intel MKL can only run on x86/x64 machines and is the main blocker for expanding to other architectures.
```

The current wording makes it sound like all of ML.NET needs MKL. But in reality only one trainer and 2 transforms do (IIRC).

Author: Let me figure out the exact trainers/transforms that do. Even if its only 2 of each I would still say its a hard dependency for 2 main reasons:
 - The build process itself requires it. If Intel MKL is not present, the native components of the build will fail.
 - The base test class tries to setup Intel MKL in its constructor, even for tests that don't use it at all.

With these 2 things, you can't build or test without Intel MKL present, even if you are not using anything that needs it.

Thoughts?

ML.NET also has dependencies on things that either don't build on other architectures or have to be compiled by the user if its wanted. For example:
 - LightGBM
 - TensorFlow
 - OnnxRuntime

I will go over these in more depth below.

### 2.2 Build
Since ML.NET has a hard dependency on Intel MKL, the build process assumes it's always there. For example, the build process will try and copy dlls without checking if they exist. The build process will need to be modified so that it doesn't fail when it can't find these files. It does the same copy check for our own Native dlls, so this will need to be fixed for those as well.

### 2.3 Managed Code

> Since ML.NET has a hard dependency on Intel MKL, the managed code imports the dlls without checking whether or not they exist. If the dlls don't exist you get a hard failure. For example, the base test class imports and sets up Intel MKL even if the test itself does not need it. This same situation applies to our own native dlls. When they aren't present, the imports fail. We will need to make ML.NET correctly handle the dll imports.

Reviewer: You mention tests here, what about the product?

Author: The product does this as well. I was just trying to give an example of how the current behavior is regarding Intel MKL and native dependencies. Do you think I should make it more clear that it does in the product as well?

Reviewer: Yes, the product problem is the first order issue.  The fact that the product does this means that our software fallback paths are somewhat "dead code" right now, so we there may be some additional cost involved in enabling them to ensure they function correctly.

Reviewer 2: We intentionally factored out any components that depend on MKL into the `Microsoft.ML.Mkl.Components` library. It makes sense to me that since these components are in "MKL.Components" that they assume MKL is available.

### 2.4 Native Projects

> ML.NET has 6 native projects. They are:

Reviewer: The below are good notes on the findings.  To help compare/contrast can you classify them consistently:
1. Has managed fallback already?
2. Reason for native use, eg: perf, use of native dependency, historical/arbitrary, etc.
3. Platform port cost (T-shirt)
4. Managed port cost (T-shirt) and implications (eg: poor perf, complexity, etc)

This can directly feed into recommendations for solutions.

 - CpuMathNative

>    - CpuMathNative relies heavily on IntelMKL. Using NetCore 3.1, however, there is a full software fallback.

Reviewer 2: ```suggestion
   - CpuMathNative relies heavily on x86 SIMD operators. Using NetCore 3.1, however, there is a full software fallback.
```

CpuMathNative doesn't reference MKL

Author: Thanks for the clarification. I thought that the x86 SIMD operators were part of Intel MKL itself. Can we sync more offline so I can better understand if/how this impacts anything?

 - FastTreeNative

>    - FastTreeNative has a software fallback using a C# compiler flag. This flag is hardcoded to always use the native code.

Reviewer 2: What about FastTreeNative doesn't compile for arm? My assumption was that this one might just work.

Author: `segment.cpp` uses the x86 SIMD operators.

 - LdaNative
   - Builds successfully without native dependencies.
 - MatrixFactorizationNative

>    - Uses libmf. Currently, we are hardcoding the "USESSE" flag which requires Intel MKL. Removing this flag allows MatrixFactorizationNative to build for Arm.

Reviewer 2: ```suggestion
   - Uses libmf. Currently, we are hardcoding the "USESSE" flag which requires an x86/x64 processor. Removing this flag allows MatrixFactorizationNative to build for Arm.
```

 - MklProxyNative
   - Wrapper for Intel MKL. When Intel MKL is not present, this is not needed. However, the build is hardcoded to always compile this code.
 - SymSgdNative
   - Only uses 4 Intel MKL methods. For 2 of them, I can't find direct replacements on other architectures.

Of these 6, only LdaNative builds successfully for other architectures without changing anything.

### 2.5 3rd Party Dependencies
As mentioned above, there are several 3rd party packages that don't have support for non x86/x64 machines.
 - LightGBM. LightGBM doesn't offer packages for non x86/x64. I was able to build the code for Arm64, but we would either have to build it ourselves, convince them to create more packages for us, or annotate that this doesn't work on non x86/64 machines.
 - TensorFlow. The full version of TensorFlow only runs on x86/x64. There is a [lite](https://www.tensorflow.org/lite/guide/build_arm64) version that supports Arm64, and you can install it directly with python, but this isn't the full version so not all models will run. We would also have to verify if the C# library we use to interface with TensorFlow will work with the lite version.

>  - OnnxRuntime. OnnxRuntime doesn't have prebuilt packages for more than just x86/x64. It does support Arm, but we have to build it ourselves. This is the same situation as with LightGBM.

Reviewer 2: ```suggestion
 - OnnxRuntime. OnnxRuntime doesn't have prebuilt packages for more than just x86/x64. It does support Arm, but we have to build it ourselves or get the OnnxRuntime team to package Arm assemblies. This is the same situation as with LightGBM.
```

## 3 Possible Solutions
We have several possible options we can use to resolve this:

>  - Fully remove Intel MKL as a dependency and use software that supports all platforms/architectures.

Reviewer: Rather than focus on Intel MKL here, can we rephrase to just say something like
 - Keep ML.NET native components and rewrite to a avoid problematic dependencies completely
 - Keep ML.NET native components and ifdef/rewrite to avoid problematic dependencies on platforms/architectures only where they are not supported.
 - Eliminate ML.NET native components and implement all functionality in managed code.
 - Hybrid

>  - Fully remove Intel MKL as a dependency and use software that supports all platforms/architectures.

Reviewer 2: Note that we did some investigation on replacing MKL with Math.NET, and the perf difference is pretty drastic.

https://github.com/dotnet/machinelearning/pull/1392#issuecomment-436888201

Author: Do you know if Math.NET is using Hardware intrinsics?

Reviewer 2: I don't think it does:

https://github.com/mathnet/mathnet-numerics/issues/565

 - Fix the build so it's not hardcoded to look for Intel MKL or any native binaries.
 - Fix the managed code so that if it can't find Intel MKL or any native binaries, the code behaves correctly. This includes software fallbacks and/or description error messages.
 - CpuMathNative has software fallbacks already in place for .NET core 3.1, so no work is needed.
 - FastTreeNative has a flag for a software fallback. We would need to conditionally enable this. Alternatively, we could change the C# code so the software fallback is always enabled for the cases when it can't find the native binaries.
 - LdaNative builds just fine for Arm, so no work would be required.
 - MatrixFactorizationNative builds with the "USESSE" flag that requires Intel MKL. We can conditionally enable/disable that flag and no other work would be required.
 - MklProxyNative can be removed/ignored for Arm builds. We will need to modify the build to exclude it as needed.
 - SymSgdNative needs to be either re-written in managed code, or re-write 4 Intel MKL methods. The 4 methods are just dealing with vector manipulation and shouldn't be hard to do.

## My Suggestion

> My suggestion would be to start with the hybrid approach. It will require the least amount of work to get ML.NET running elsewhere, while still being able to support a large majority of devices out of the gate. This solution will still limit the platforms we can run-on to Arm64 devices, but we can do a generic Arm64 compile, so it should work for all Arm64 v8 devices. It would be a good idea to have full software fallbacks so that truly anywhere .NET Core runs, ML.NET will run, such as Web Assembly and when .NET 6 comes out on mobile as well.

Reviewer 3: Once a fallback is in place, I might recommend benchmarking the speed gain MKL gives us. 

If the speed gain is not significant, removing MLK will give the benefit of simplicity. Historically, there has been many filed [issues related to MLK](https://github.com/dotnet/machinelearning/search?q=MKL&type=issues) (build issues, runtime issues, maintenance issues).

### NET Core 5
I think we should also add support for .NET Core 5 during this process so that we gain access to the Arm64 intrinsics. However, since .Net Core 5 goes out of support before .NET core 3.1 I don't think .NET Core 5 should be a huge focus.

### 3rd Party Dependencies
I think initially we should annotate that they don't work on non x86/x64 devices. This includes logging an error when they try and run an unsupported 3rd party dependency, and then failing gracefully with a helpful and descriptive error. The user should be able to compile the 3rd party dependency, for the ones that support it, and have ML.NET still be able to pick it up and run it if it exists. OnnxRuntime is something that we will probably want, but we can look more into this as we get requests for it in the future.

### Helix

> In order to fully test everything we need to, we would also need to change how we test to use the Helix testing servers. Currently, Helix doesn't have the capability to test Apple's new M1 code, but that is in the works.

Author: Lets sync on this tomorrow.

> In order to fully test everything we need to, we would also need to change how we test to use the Helix testing servers. Currently, Helix doesn't have the capability to test Apple's new M1 code, but that is in the works.

Author: Helix section has been updated.

### Mobile Support

> .NET Core 6 will allow us to run nativly on mobile. Since we are making these changes before .NET 6 is released, I propose we don't include that work as of yet. As long as we handle the native binaries correctly and make sure ML.NET provides descriptive error methods, we should be able to have mobile support as soon as .NET Core 6 releases for everything that currently has a software fallback.

Reviewer: It's important to know what we intend to do here.  Do we think we'll be able to make all our native components work on these new platforms?  If not, wouldn't that change how we're thinking about which components should get a managed fallback?

Author: Updated a bit, but does mobile have issues with loading native modules?

Reviewer: WASM doesn't today.  iOS and Android do, but would require explicit targeting and not all dependencies may be present.  There are also potentially code execution constraints on these platforms (iOS -> no runtime codegen).

### Support Grid

> This is what I propose for the support grid. Since .NET Core 2.1 is end-of-life this year, I am putting much less emphasis on it. Since .NET Core 5 will be out of support before .NET Core 3.1 will be, I am putting less CI emphasis on .NET Core 5.

Reviewer: Can we split up the statements around "support" and "test" (perhaps build as well).  Support is a statement we make to customers.  Build and test is how we achieve that.  I think it's reasonable to have a sparse matrix for testing where we focus on LTS releases, while still achieving coverage of our built assets.

Author: Lets sync on this tomorrow.

> | Platform | Architecture | Intel MKL | .NET Framework | .NET Core 2.1 | .NET Core 3.1 | .NET Core 5 |

Reviewer: Which versions and distributions of the below are we going to support?

Author: Lets sync on this tomorrow.

| ---------| -------------| --------- | -------------- | ------------- | ------------- | ----------- |
| Windows  | x64          | Yes       | Yes            | Yes, no CI    | Yes           | Yes         |

> | Windows  | x86          | Yes       | Yes, no CI     | Yes, no CI    | Yes, no CI    | Yes, no CI  |

Reviewer 2: Why not Windows Arm64?

Author: I actually had that thought as well and meant to add it before I made the PR. Obviously I forgot, I'll add it.

| Mac      | x64          | Yes       | No             | Yes, no CI    | Yes           | Yes, no CI  |
| Mac      | Arm64        | No        | No             | No            | Yes           | Yes, no CI  |

> | Ios      | Arm64        | No        | No             | No            | No            | No          |

Reviewer 2: I believe iOS x64 is needed in order to run under a simulator. But I could be wrong - we should verify.

Author: What do you mean under a simulator? I know there is `MacCatalyst`, which lets you run IOs apps on the Mac itself, is that what you are meaning?

Reviewer 2: https://developer.apple.com/documentation/xcode/running_your_app_in_the_simulator_or_on_a_device

		- [2.5 3rd Party Dependencies](#25-3rd-party-dependencies)
	- [3 Possible Solutions](#3-possible-solutions)
		- [3.1 Rewrite native components to work on other platforms](#31-rewrite-native-components-to-work-on-other-platforms)
		- [3.2 Rewrite native components to be only managed code](#32-rewrite-native-components-to-be-only-managed-code)
		- [3.3 Hybrid Between Finding Replacements and Software Fallbacks](#33-hybrid-between-finding-replacements-and-software-fallbacks)
	- [My Suggestion](#my-suggestion)
		- [Improving managed fallback through intrinsics](#improving-managed-fallback-through-intrinsics)
		- [3rd Party Dependencies](#3rd-party-dependencies)
		- [Helix](#helix)
		- [Mobile Support](#mobile-support)
		- [Support Grid](#support-grid)

## 1. Why cross-platform/architecture
ML.NET is an open-source machine learning framework which makes machine learning accessible to .NET developers with the same code that powers machine learning across many Microsoft products, including Power BI, Windows Defender, and Azure.

ML.NET allows .NET developers to develop/train their own models and infuse custom machine learning into their applications using .NET.

> Currently, while .NET is able to run on many different platforms and architectures, ML.NET is only able to run on Windows, Mac, and Linux, either x86 or x64. This excludes many architectures such as Arm, Arm64, M1, and web assembly, places that .NET is currently able to run.

Reviewer 4: ```suggestion
Currently, while .NET is able to run on many different platforms and architectures, ML.NET is only able to run on Windows, Mac, and Linux, either x86 or x64. This excludes many architectures such as Arm, Arm64, Apple Silicon, and web assembly (WASM), places that .NET is currently able to run.
```

> Currently, while .NET is able to run on many different platforms and architectures, ML.NET is only able to run on Windows, Mac, and Linux, either x86 or x64. This excludes many architectures such as Arm, Arm64, M1, and web assembly, places that .NET is currently able to run.

Reviewer 4: I would say the most important platform is Arm64/Apple Silicon followed by WASM followed by Arm. So if something is going to be slow on Arm (after all we don't support intrinsics there), or possibly even missing, that shouldn't necessarily prevent us doing the right thing for Arm64.

Author: From what I have seen and tested so far, arm vs arm64 should mostly take the same work to get both working. The main differences should mostly be with the intrinsics, if we add them. That may change as I get farther into the implementation, but that seems to be the case currently.

> The goal is to enable ML.NET to run everywhere that .NET itself is able to run.

Reviewer 4: Are there secondary goals? Eg 
1. Do not regress ML.NET performance on platforms where it currently runs. 
2. Where possible, aim for convergence to a single implementation for all platforms. (These goals may be in conflict..)
3. Where possible, avoid taking on ownership of code that is not core to ML.NET or does not align with our expertise. For example, we would probably not want to maintain a large math library.

Reviewer 4: Another: 
4. Optimize for progressive delivery to reduce risk. ie., a solution where we expand to one platform at a time, or with partial functionality, release, and continue. No big bang effort that succeds or fails together.

Reviewer 4: 5. Where possible, prefer that dependencies be supplied with the platform (possibly a package manager) rather than shipped as part of ML.NET. (this is a weaker goal, as usability is important, so maybe it's a non goal)

## 2. Current status
There are several problems complicating us from moving to a fully cross-platform solution. At a high level these include:
1. We only build and test for a subset of platforms that .NET supports.
2. Native components must be explicitly built for additional architectures which we wish to support. This limits our ability to support new platforms without doing work.
3. Building our native components for new platforms faces challenges due to lack of support for those components dependencies. This limits our ability to support the current set of platforms.
4. Some of our external dependencies have limited support for .NET's supported platforms.

> 5. Some things we use internally are optomized for x86/x64 and wont work well on other platforms. For example various components are parallelized but current webassembly targets are single threaded. It's likely some changes will be necessary to various algorithms to work well in these environments.

Reviewer 4: ```suggestion
5. Some things we use internally are optimized for x86/x64 and wont work well on other platforms. For example various components are parallelized but current web assembly targets are single threaded. It's likely some changes will be necessary to various algorithms to work well in these environments.
```

> ### 2.1 Problems

Reviewer 5: What about other problems? Does it work on platforms with memory/cpu limitations? Does it use any APIs which is not universally available (e.g threads) ?

 - LightGBM
 - TensorFlow
 - ONNX Runtime

I will go over these in more depth below.

### 2.2 Build
Since ML.NET has a hard dependency on x86/x64, the build process assumes it's running there. For example, the build process will try and copy native DLLs without checking if they exist because it assumes the build for them succeeded or that they are available. The build process will need to be modified so that it doesn't fail when it can't find these files. It does the same copy for our own Native DLLs, so this will need to be fixed for those as well.

### 2.3 Managed Code
Since ML.NET has a hard dependency on x86/x64, the managed code imports DLLs without checking whether or not they exist. If the DLLs don't exist you get a hard failure. For example, if certain columns are active, the `MulticlassClassificationScorer` will call `CalculateIntermediateVariablesNative` which is loaded from `CpuMathNative`, but all of this is done without any checks to see if the DLL actually exists. The tests also run into this problem, for instance, the base test class imports and sets up Intel MKL even if the test itself does not need it.

> ### 2.4 Native Projects

Reviewer 4: Do any of these have transitive dependencies that might be a problem?

Author: The native projects? I don't believe there are any other dependencies or issues then the one I have listed below. As I get further into the implementation that may change, but from the testing I have done so far that seems to be the case.

ML.NET has 6 native projects. They are:
 - CpuMathNative

>    - Partial managed fallback when using NetCore 3.1.

Reviewer 4: ```suggestion
   - Partial managed fallback when using .NET Core 3.1 or later.
```
Note that .NET Core 2.1 [goes out of support on Aug 21 this year](https://dotnet.microsoft.com/platform/support/policy/dotnet-core). So Support for .NET Core 3.1+ means all .NET Core at that point.

>    - A large amount of work would be required to port the native code to other platforms. We would have to change all the SIMD instructions for each platform.

Reviewer 6: This might be partially automatable. The C++ hardware intrinsics have a largely 1-to-1 mapping with the corresponding .NET hardware intrinsics, just with different names.

Author: Is there software we have that does that automation automatically? Or would I have to create that automation as well? Is this true for arm intrinsics as well?

   - A small amount of work required for a full managed fallback.

>    - This was created before we had hardware intrinsics so we used native code for performance. The managed fallback uses x86/x64 intrinsics if the hardware supports it, otherwise it has a plain managed fallback without intrinsics if needed.

Reviewer 4: Eventually we will want to add codepaths for Arm64 intrinsics as well. They didn't exist when this work was done, but were added in .NET 5.0. That is a performance optimization, so it's not necessary for your porting effort, but it may be necessary at some point.

Reviewer 6: This may be simpler with my https://github.com/dotnet/runtime/issues/49397 proposal. This would allow a good amount of code to be written once and shared between x86 and ARM64 while operating on fixed-sized data (where `Vector<T>` is variable sized and can introduce certain problems).

You would then fallback to platform specific APIs only where an optimization exists or where a common path isn't possible.

 - FastTreeNative
   - Full managed fallback by changing a C# compiler flag. This flag is hardcoded to always use the native code.
   - Small amount of work required to change build process and verify it's correct.
   - This was created before we had hardware intrinsics so we used native code for performance. The managed fallback does not currently use hardware intrinsics, so this will be slower than the native solution. Hardware intrinsics could be added to improve performance.
 - LdaNative
   - No managed fallback, but builds successfully on non x86/x64 without changes.

>    - Large amount of work to have a managed fallback. This solution is about 3000 lines of code not including any code from the dependencies.

Reviewer 4: Question for this and the other libraries where we might have to write a bunch of managed code -- how good are the tests? If there are poor tests, that will add risk/cost because they will have to be improved before major work.

Author: So the native code has no testing itself, but the managed code that calls the native code has good/decent test coverage. Enough that if we do something wrong during this process it should be caught. Or at least enough to catch it in the majority of the cases. Overall we have about 70% test coverage, so it should be enough to catch the majority of them.

   - Small amount of work required to change the build process to exclude this as needed.
 - SymSgdNative
   - No managed fallback.
   - Medium amount of work required to have a managed fallback. Only about 500 lines of code plus having to implement 4 vector operations.
   - Small amount of work required port to other platforms. Only uses 4 Intel MKL methods which we can replace. I was only able to find 2 direct replacements on other architectures, but we could easily write our own for either those 2 or even all 4.
   - This was created before we had hardware intrinsics so we used native code for performance as well as take advantage of IntelMKL.

Of these 6, only LdaNative builds successfully for other architectures without changing anything.

### 2.5 3rd Party Dependencies
As mentioned above, there are several 3rd party packages that don't have support for non x86/x64 machines.
 - LightGBM. LightGBM doesn't offer packages for non x86/x64. I was able to build the code for Arm64, but we would either have to build it ourselves, convince them to create more packages for us, or annotate that this doesn't work on non x86/64 machines.

>  - TensorFlow. The full version of TensorFlow only runs on x86/x64. There is a [lite](https://www.tensorflow.org/lite/guide/build_arm64) version that supports Arm64, and you can install it directly with python, but this isn't the full version so not all models will run. We would also have to verify if the C# library we use to interface with TensorFlow will work with the lite version.

Reviewer 4: When non-.NET apps do inference on Android or IPhone, do they use this lite Tensorflow - that would suggest it's sufficient?

Reviewer 7: For mobile, you can use either the TFLite API on the respective platforms or things like MLKit / Core ML. The minimum requirement though would be the TFLite API.

Author: I believe that TFLite will work out of the gate with the bindings that we have. I haven't done a swap to test it for sure though, I'll do that tomorrow.

>  - ONNX Runtime. ONNX Runtime doesn't have prebuilt packages for more than just x86/x64. It does support Arm, but we have to build it ourselves or get the ONNX Runtime team to package Arm assemblies. This is the same situation as with LightGBM.

Reviewer 4: If we will need ONNX runtime and LightGBM for Arm* I suggest to reach out to them informally to see whether this is something they'd contemplate, whether it's part of their roadmap already, etc.

Author: We spoke with ONNX Runtime about it and they basically said that since there was no customer asks for it on arm that it wasn't on their soon to do list. Hopefully when ML.NET officially supports arm we can get the data needed to make that ask.

I haven't spoken with LightGBM about it. Do you have a contact for them? If not I can find something.

## 3 Possible Solutions
We have several possible options we can use to resolve this:
 - Keep ML.NET native components and rewrite them to avoid problematic dependencies completely.
 - Keep ML.NET native components and ifdef/rewrite to avoid problematic dependencies on platforms/architectures only where they are not supported.
 - Hybrid approach of replacement code and software fallback. This is the approach I recommend.

None of these approaches resolve the 3rd party dependency issues. These solutions only deal with first party ML.NET code itself.

I have lots more info about our dependency on x86/x64 in another document if required.

### 3.1 Rewrite native components to work on other platforms

> This will still allow us to gain the benefits of the X86/x64 SIMD instructions and Intel MKL on architectures that support it but will also keep the benefits of native code in the other places. The downside is that we would have to build the native code for, potentially, a lot of different architectures.

Reviewer 4: Or convince upstream to build it?

Author: Here I was meaning the native ML.NET components we wrote ourselves. No upstream in this case sadly.

 - Fix the managed code so that if it can't find the native components the code behaves correctly.
 - All the native code would need to have the x86/x64 dependencies change to other libraries or be re-written by us.
   - CpuMathNative would need to be re-written to not use x86/x64 dependencies.
   - FastTreeNative would need to be re-written to not use x86/x64 dependencies.
   - LdaNative builds just find without x86/x64 dependencies, so no replacements would need to be found.
   - MatrixFactorizationNative builds with the "USESSE" flag that requires x86/x64. We can conditionally enable/disable that flag and no other work would be required.
   - MklProxyNative is only a wrapper for Intel MKL and can be ignored on non x86/x64 platforms. We will need to modify the build to exclude it as needed.
   - SymSgdNative would need to be re-written to not use Intel MKL. There are only 4 methods that would need to be changed.

I was unable to find all the replacements we would need. We would end up having to write many native methods ourselves with this approach.

### 3.2 Rewrite native components to be only managed code

> This will truly allow ML.NET to run anywhere that .NET runs. The only downside is the speed of execution, and the time to rewrite the existing native code we have.  If we restrict new architectures to .NET core 3.1 or newer, we will have an easier time with the software fallbacks as some of this code has already been written. This solution will also require a lot of code rewrite from native code to managed code.

Reviewer 4: Aside from Arm, where we do not have intrinsics, I would not assume that the speed need be signfiicantly compromised. We have had good results, eg., moving parts of native code of CoreCLR into managed. However, it needs careful optimization, and sometimes JIT work. So I can't claim it's cheap - but it's no longer necessarily slower

I suppose if the native code uses special tricks like GPU then we can't compete.

Author: That is a good point. Let me rephrase that to better state that in the doc. It would only be slower if we didn't do all those steps you mentioned.

> This will truly allow ML.NET to run anywhere that .NET runs. The only downside is the speed of execution, and the time to rewrite the existing native code we have.  If we restrict new architectures to .NET core 3.1 or newer, we will have an easier time with the software fallbacks as some of this code has already been written. This solution will also require a lot of code rewrite from native code to managed code.

Reviewer 7: I believe with .NET Core 3.1 you have issues with WebAssembly because it looks for CpuMathNative.

Author: It looks for it but its not needed. Thats part of the build process itself. I will be fixing that as part of this process.

At a high level, this solution would require us to:
 - Fix the build so it's not hardcoded to look for any native dependencies or binaries.
 - Fix the managed code so that if it can't find the native binaries, the code behaves correctly and performs the software fallback.

>  - CpuMathNative mostly has software fallbacks already in place for .NET core 3.1, so only a little work is needed.

Reviewer 4: ```suggestion
 - CpuMathNative mostly has software fallbacks already in place for .NET Core 3.1 and later, so only a little work is needed.
```

 - FastTreeNative has a flag for a software fallback. We would need to conditionally enable this. Alternatively, we could change the C# code so the software fallback is always enabled for the cases when it can't find the native binaries.
 - LdaNative would need to be re-written.
 - MatrixFactorizationNative would need to be re-written.
 - MklProxyNative can be ignored. It is not needed with software fallbacks. We will need to modify the build to exclude it as needed.
 - SymSgdNative would need to be re-written.

### 3.3 Hybrid Between Finding Replacements and Software Fallbacks
Since some of the native code already has replacements and some of the code already has software fallbacks, we can leverage this work by doing a hybrid between the prior 2 solutions.

At a high level, this solution would require us to:
 - Fix the build so it's not hardcoded to look for any native dependencies or binaries.
 - Fix the managed code so that if it can't find the native binaries, the code behaves correctly and performs the software fallback. This includes software fallbacks and/or description error messages.

>  - CpuMathNative mostly has software fallbacks already in place for .NET core 3.1, so only a little work is needed.

Reviewer 4: ```suggestion
 - CpuMathNative mostly has software fallbacks already in place for .NET Core 3.1 and later, so only a little work is needed.
```

 - FastTreeNative has a flag for a software fallback. We would need to conditionally enable this. Alternatively, we could change the C# code so the software fallback is always enabled for the cases when it can't find the native binaries.
 - LdaNative builds just find without x86/x64 dependencies, so no replacements would need to be found.
 - MatrixFactorizationNative builds with the "USESSE" flag that requires x86/x64. We can conditionally enable/disable that flag and no other work would be required.
 - MklProxyNative is only a wrapper for Intel MKL and can be ignored on non x86/x64 platforms. We will need to modify the build to exclude it as needed.
 - SymSgdNative needs to be either re-written in managed code, or re-write 4 Intel MKL methods. The 4 methods are just dealing with vector manipulation and shouldn't be hard to do.

## My Suggestion

> My suggestion would be to start with the hybrid approach. It will require the least amount of work to get ML.NET running elsewhere, while still being able to support a large majority of devices out of the gate. This solution will still limit the platforms we can run-on to what we build the native components for, initially Arm64 devices, but we can do a generic Arm64 compile so it should work for all Arm64 v8 devices. The goal is to eventually have a managed implementation which can work everywhere .NET does and accelerated components to increase performance where possible. This could include native components in some cases or hardware intrinsics in others.

Reviewer 4: sounds good to me

> My suggestion would be to start with the hybrid approach. It will require the least amount of work to get ML.NET running elsewhere, while still being able to support a large majority of devices out of the gate. This solution will still limit the platforms we can run-on to what we build the native components for, initially Arm64 devices, but we can do a generic Arm64 compile so it should work for all Arm64 v8 devices. The goal is to eventually have a managed implementation which can work everywhere .NET does and accelerated components to increase performance where possible. This could include native components in some cases or hardware intrinsics in others.

Reviewer 7: 👍

### Improving managed fallback through intrinsics

> We should also target .NET 5 so that we gain access to the Arm64 intrinsics. Rather than implementing special-purpose native libraries to take advantage of architecture-specific instructions we should instead enhance performance ensuring our managed implementation leverage intrinsics.

Reviewer 4: It may well be OK if Arm* only works on .NET 5+, or it's slow on 3.1 only.

Reviewer 8: I was about to write up something similar in the Support Grid section.

If supporting only .NET Core 5+ saves us a lot of effort or complexity. Perhaps this would be a good time to bump the major version of ML.NET and reduce the number of .NET versions we support?

Author: We won't save a lot of effort/complexity just by going to .NET 5. We could gain execution speed if we switched to .NET 5+ and also implemented the intrinsics for it. It may be worth it to update the major version of ML.NET once we have all those hardware intrinsics in place for arm.

### 3rd Party Dependencies
I think initially we should annotate that they don't work on non x86/x64 devices. This includes logging an error when they try and run an unsupported 3rd party dependency, and then failing gracefully with a helpful and descriptive error. The user should be able to compile the 3rd party dependency, for the ones that support it, and have ML.NET still be able to pick it up and run it if it exists. ONNX Runtime is something that we will probably want, but we can look more into this as we get requests for it in the future.

### Helix

> In order to fully test everything we need to, we would also need to change how we test to use the Helix testing servers. Currently, Helix doesn't have the capability to test Apple's new M1 code, but that is in the works. We will be building once for each architecture/platform combination, and then fan out and submit one job for each framework version we want to test. We will also be cross-targeting builds. For example we can build on normal linux to target arm linux and then run those tests using Helix. Its estimated to be about a medium amount of work to make the changes required to use Helix.

Reviewer 5: > Helix doesn't have the capability to test Apple's new M1 code

I don't think that accurate, we have been using it for some time

### Mobile Support

> .NET Core 6 will allow us to run natively on mobile. Since we are making these changes before .NET 6 is released, I propose we don't include that work as of yet. As long as we handle the native binaries correctly and make sure ML.NET provides descriptive error methods, we should be able to have mobile support as soon as .NET Core 6 releases for everything that currently has a software fallback. Since the native projects we are proposing to keep build for Arm64, they should work on mobile as well.

Reviewer 4: > Since the native projects we are proposing to keep build for Arm64, they should work on mobile as well.

This is possibly not the case depending on whether they have OS/platform dependencies.

Author: Thats good to know. I will need to look more into this. Thanks for the heads up.

Reviewer 5: If you have native dependencies you will have to build them differently for mobile platforms.  The arm64 aspect is only one angle, you need to deal with different libc, etc as well

> .NET Core 6 will allow us to run natively on mobile. Since we are making these changes before .NET 6 is released, I propose we don't include that work as of yet. As long as we handle the native binaries correctly and make sure ML.NET provides descriptive error methods, we should be able to have mobile support as soon as .NET Core 6 releases for everything that currently has a software fallback. Since the native projects we are proposing to keep build for Arm64, they should work on mobile as well.

Reviewer 7: To make sure I understand correctly, by making the proposed changes, once .NET 6 is out ML.NET is expected to work out of the box on those platforms. The table below refers to CI efforts.

### Support Grid

> This is what I propose for the support grid. Since .NET Core 2.1 is end-of-life this year, I am putting much less emphasis on it. Since .NET Core 5 will be out of support before .NET Core 3.1 will be, I am putting less CI emphasis on .NET Core 5.

Reviewer 4: As above, I would do zero 2.1-specific coding effort. Maybe just cursory testing.

Author: I haven't heard that phrase before. Can you explain in more depth?

Reviewer: I think Dan is just mentioning that we don't invest in making our .NETCoreApp2.1 builds work on more platforms, be faster, or have more test coverage.   Since the platform is going out of support soon and it should be "easier" to do work on newer frameworks, then we should focus our efforts on the newer frameworks.

| Platform | Architecture | Intel MKL | .NET Framework | .NET Core 2.1 | .NET Core 3.1 | .NET Core 5 |
| ---------| -------------| --------- | -------------- | ------------- | ------------- | ----------- |
| Windows  | x64          | Yes       | Yes            | Yes, no CI    | Yes           | Yes         |
| Windows  | x86          | Yes       | Yes, no CI     | Yes, no CI    | Yes, no CI    | Yes, no CI  |
| Mac      | x64          | Yes       | No             | Yes, no CI    | Yes           | Yes, no CI  |
| Mac      | Arm64        | No        | No             | No            | Yes           | Yes, no CI  |

> | Ios      | Arm64        | No        | No             | No            | No            | No          |

Reviewer 5: iOS is the correct name for the platform

> | Ios      | x64          | No        | No             | No            | No            | No          |

Reviewer 5: Note: .NET6 will also support iOS on arm

	- [My Suggestion](#my-suggestion)
		- [Improving managed fallback through intrinsics](#improving-managed-fallback-through-intrinsics)
		- [3rd Party Dependencies](#3rd-party-dependencies)
		- [Helix](#helix)
		- [Mobile Support](#mobile-support)
		- [Support Grid](#support-grid)

## 1. Why cross-platform/architecture
ML.NET is an open-source machine learning framework which makes machine learning accessible to .NET developers with the same code that powers machine learning across many Microsoft products, including Power BI, Windows Defender, and Azure.

ML.NET allows .NET developers to develop/train their own models and infuse custom machine learning into their applications using .NET.

Currently, while .NET is able to run on many different platforms and architectures, ML.NET is only able to run on Windows, Mac, and Linux, either x86 or x64. This excludes many architectures such as Arm, Arm64, M1, and web assembly, places that .NET is currently able to run.

The goal is to enable ML.NET to run everywhere that .NET itself is able to run.

## 2. Current status

> There are several problems complicating us from moving to a fully cross-platform solution. At a high level thes include:

Reviewer 3: Nit grammar:
```suggestion
There are several problems complicating us from moving to a fully cross-platform solution. At a high level these include:
```

1. We only build and test for a subset of platforms that .NET supports.
2. Native components must be explicitly built for additional architectures which we wish to support. This limits our ability to support new platforms without doing work.
3. Building our native components for new platforms faces challenges due to lack of support for those components dependencies. This limits our ability to support the current set of platforms.

> 4. Some of our external dependencies have limited support for .NET's supported platforms.

Reviewer 9: There are also some things ML.NET uses internally that don't work well on some platforms, e.g. various components are parallelized but current webassembly targets are single threaded.  It's likely some changes will be necessary to various algorithms to work well in these environments.

Author: That is a good point I hadn't though about. Let me add that point as well.

### 2.1 Problems
ML.NET has a hard dependency on x86/x64. Some of the dependency is on Intel MKL, while other parts depend on x86/x64 SIMD instructions. To make things easier I will refer to these as just the x86/x64 dependencies. This is to perform many optimized math functions and enables several transformers/trainers to be run natively for improved performance. The problem is that these dependencies can only run on x86/x64 machines and are the main blockers for expanding to other architectures. While you can run the managed code on other architectures, there is no good way to know which parts will run and which ones wont. This includes the build process as well, which currently has these same hard dependencies and building on non x86/x64 machines is not supported.

> ML.NET also has dependencies on things that either don't build on other architectures or have to be compiled by the user if its wanted. For example:

Reviewer 3: Nit grammar: 
```suggestion
ML.NET also has dependencies on things that either don't build on other architectures or have to be compiled by the user if it's wanted. For example:
```

 - LightGBM
 - TensorFlow

>  - OnnxRuntime

Reviewer 3: Stylize to match https://github.com/microsoft/onnxruntime:
```suggestion
 - ONNX Runtime
```

I will go over these in more depth below.

### 2.2 Build

> Since ML.NET has a hard dependency on x86/x64, the build process assumes it's running there. For example, the build process will try and copy native dlls without checking if they exist because it assumes the build for them succeeded or that they are available. The build process will need to be modified so that it doesn't fail when it can't find these files. It does the same copy for our own Native dlls, so this will need to be fixed for those as well.
> 
> ### 2.3 Managed Code
> Since ML.NET has a hard dependency on x86/x64, the managed code imports dlls without checking whether or not they exist. If the dlls don't exist you get a hard failure. For example, if certain columns are active the `MulticlassClassificationScorer` will call `CalculateIntermediateVariablesNative` which is loaded from `CpuMathNative`, but all of this is done without any checks to see if the DLL actually exists. The tests also run into this probleb. The base test class imports and sets up Intel MKL even if the test itself does not need it.

Reviewer 3: Nit--copy editing: _(DLL to match Wikipedia capitalization, and misspelling of problem_)
```suggestion
Since ML.NET has a hard dependency on x86/x64, the build process assumes it's running there. For example, the build process will try and copy native DLLs without checking if they exist because it assumes the build for them succeeded or that they are available. The build process will need to be modified so that it doesn't fail when it can't find these files. It does the same copy for our own Native DLLs, so this will need to be fixed for those as well.

### 2.3 Managed Code
Since ML.NET has a hard dependency on x86/x64, the managed code imports DLLs without checking whether or not they exist. If the DLLs don't exist you get a hard failure. For example, if certain columns are active, the `MulticlassClassificationScorer` will call `CalculateIntermediateVariablesNative` which is loaded from `CpuMathNative`, but all of this is done without any checks to see if the DLL actually exists. The tests also run into this problem, for instance, the base test class imports and sets up Intel MKL even if the test itself does not need it.
```

Author: Good catch.

### 2.4 Native Projects
ML.NET has 6 native projects. They are:
 - CpuMathNative
   - Partial managed fallback when using NetCore 3.1.
   - A large amount of work would be required to port the native code to other platforms. We would have to change all the SIMD instructions for each platform.
   - A small amount of work required for a full managed fallback.

>    - This was created before we had hardware intrinsics so we used native code for performance. The managed fallback uses x86/x64 intrinsics where possible.

Reviewer 9: > This was created before we had hardware intrinsics so we used native code for performance. The managed fallback uses x86/x64 intrinsics where possible.

Does the native code utilize functionality not available now via intrinsics?  I'm wondering about the "where possible" distinction.

Author: What I was trying to mean is that it uses AVX intrinsics if AVX is supported, if its not it tries SSE intrinsics, and finally it just has a non-intrinsics fallback. So by "where possible" I was meaning possible based on the hardware its running on. Let me add this detail in.

 - FastTreeNative
   - Full managed fallback by changing a C# compiler flag. This flag is hardcoded to always use the native code.
   - Small amount of work required to change build process and verify it's correct.

>    - This was created before we had hardware intrinsics so we used native code for performance. The managed fallback does not use hardware intrinsics, so this will be slower than the native solution.

Reviewer 9: > The managed fallback does not use hardware intrinsics

But could with work, presumably?

Author: Yeah. Let me change the wording to say that the current state doesn't use them but they could be added in.

 - LdaNative
   - No managed fallback, but builds successfully on non x86/x64 without changes.
   - Large amount of work to have a managed fallback. This solution is about 3000 lines of code not including any code from the dependencies.
   - This was created before we had hardware intrinsics so we used native code for performance.
 - MatrixFactorizationNative

>    - No sotware fallback.

Reviewer 9: ```suggestion
   - No software fallback.
```

Author: Thanks for the catch!

   - Currently we are hardcoding the "USESSE" flag which requires x86/x64 SIMD commands. Removing this flag allows MatrixFactorizationNative to build for other platforms.

>    - Small amount of work required to change the build process and verify its working.

Reviewer 3: Nit:
```suggestion
   - Small amount of work required to change the build process and verify it's working.
```

   - Large/Xlarge amount of work to have a managed fallback. This uses libmf, so we would have to not only port our code, but we would have to understand which parts of libmf are being used and port those as well.
   - This was created before we had hardware intrinsics so we used native code for performance as well as take advantage of libmf.
 - MklProxyNative
   - Wrapper for Intel MKL. When Intel MKL is not present, this is not needed. However, the build is hardcoded to always compile this code.
   - Small amount of work required to change the build process to exclude this as needed.
 - SymSgdNative
   - No managed fallback.

>    - Medium amount of work required to have a managed fallback. Only about 500 lines of code plus having to implmement 4 vector operations.

Reviewer 3: Nit:
```suggestion
   - Medium amount of work required to have a managed fallback. Only about 500 lines of code plus having to implement 4 vector operations.
```

   - This was created before we had hardware intrinsics so we used native code for performance as well as take advantage of IntelMKL.

Of these 6, only LdaNative builds successfully for other architectures without changing anything.

### 2.5 3rd Party Dependencies
As mentioned above, there are several 3rd party packages that don't have support for non x86/x64 machines.
 - LightGBM. LightGBM doesn't offer packages for non x86/x64. I was able to build the code for Arm64, but we would either have to build it ourselves, convince them to create more packages for us, or annotate that this doesn't work on non x86/64 machines.
 - TensorFlow. The full version of TensorFlow only runs on x86/x64. There is a [lite](https://www.tensorflow.org/lite/guide/build_arm64) version that supports Arm64, and you can install it directly with python, but this isn't the full version so not all models will run. We would also have to verify if the C# library we use to interface with TensorFlow will work with the lite version.

>  - OnnxRuntime. OnnxRuntime doesn't have prebuilt packages for more than just x86/x64. It does support Arm, but we have to build it ourselves or get the OnnxRuntime team to package Arm assemblies. This is the same situation as with LightGBM.

Reviewer 3: Nit: (match casing of https://github.com/microsoft/onnxruntime)
```suggestion
 - ONNX Runtime. ONNX Runtime doesn't have prebuilt packages for more than just x86/x64. It does support Arm, but we have to build it ourselves or get the ONNX Runtime team to package Arm assemblies. This is the same situation as with LightGBM.
```

## 3 Possible Solutions
We have several possible options we can use to resolve this:
 - Eliminate ML.NET native components and implement all functionality in managed code.

>  - Keep ML.NET native components and rewrite them to a avoid problematic dependencies completely.

Reviewer 3: Nit:
```suggestion
 - Keep ML.NET native components and rewrite them to avoid problematic dependencies completely.
```

 - Fix the build so it's not hardcoded to look for specific native dependencies.
 - Fix the managed code so that if it can't find the native components the code behaves correctly.
 - All the native code would need to have the x86/x64 dependencies change to other libraries or be re-written by us.
   - CpuMathNative would need to be re-written to not use x86/x64 dependencies.
   - FastTreeNative would need to be re-written to not use x86/x64 dependencies.
   - LdaNative builds just find without x86/x64 dependencies, so no replacements would need to be found.
   - MatrixFactorizationNative builds with the "USESSE" flag that requires x86/x64. We can conditionally enable/disable that flag and no other work would be required.
   - MklProxyNative is only a wrapper for Intel MKL and can be ignored on non x86/x64 platforms. We will need to modify the build to exclude it as needed.
   - SymSgdNative would need to be re-written to not use Intel MKL. There are only 4 methods that would need to be changed.

I was unable to find all the replacements we would need. We would end up having to write many native methods ourselves with this approach.

> ### 3.3 Rewrite native components to be only managed code

Reviewer 9: What's the difference between 3.1 (eliminate native components) and 3.3 (rewrite native components to be only managed code)?

Author: So 3.1 was written back when I was thinking more about Intel MKL and not the native code as a whole. For 3.1 I was originally meaning fully remove the dependency we have on Intel MKL, but thats not applicable anymore. Your idea with 3.3 is what I was trying to convey with 3.4. I will re-word this section. 3.1 is going to be fully removed, and I will try and make the other differences more apparent. 

I think if we redid everything with hardware intrinsics we probably wouldn't need to use the native components, but I haven't done any perf testing myself to confirm that. I can run some preliminary tests with CPUMath since we have native and intrinsics for that.

## My Suggestion
My suggestion would be to start with the hybrid approach. It will require the least amount of work to get ML.NET running elsewhere, while still being able to support a large majority of devices out of the gate. This solution will still limit the platforms we can run-on to what we build the native components for, initially Arm64 devices, but we can do a generic Arm64 compile so it should work for all Arm64 v8 devices. The goal is to eventually have a general purpose implementation which can work everywhere .NET does and accelerated components to increase performance where possible, such as running on Web Assembly and when .NET 6 comes out on mobile as well.

### Improving managed fallback through intrinsics
We should also target .NET 5 so that we gain access to the Arm64 intrinsics. Rather than implementing special-purpose native libraries to take advantage of architecture-specific instructions we should instead enhance performance ensuring our managed implementation leverage intrinsics.

### 3rd Party Dependencies

> I think initially we should annotate that they don't work on non x86/x64 devices. This includes logging an error when they try and run an unsupported 3rd party dependency, and then failing gracefully with a helpful and descriptive error. The user should be able to compile the 3rd party dependency, for the ones that support it, and have ML.NET still be able to pick it up and run it if it exists. OnnxRuntime is something that we will probably want, but we can look more into this as we get requests for it in the future.

Reviewer 3: Nit: (match casing of https://github.com/microsoft/onnxruntime)
```suggestion
I think initially we should annotate that they don't work on non x86/x64 devices. This includes logging an error when they try and run an unsupported 3rd party dependency, and then failing gracefully with a helpful and descriptive error. The user should be able to compile the 3rd party dependency, for the ones that support it, and have ML.NET still be able to pick it up and run it if it exists. ONNX Runtime is something that we will probably want, but we can look more into this as we get requests for it in the future.
```

### Helix
In order to fully test everything we need to, we would also need to change how we test to use the Helix testing servers. Currently, Helix doesn't have the capability to test Apple's new M1 code, but that is in the works.

### Mobile Support

> .NET Core 6 will allow us to run nativly on mobile. Since we are making these changes before .NET 6 is released, I propose we don't include that work as of yet. As long as we handle the native binaries correctly and make sure ML.NET provides descriptive error methods, we should be able to have mobile support as soon as .NET Core 6 releases for everything that currently has a software fallback. Since the native projects we are proposing to keep build for Arm64, they should work on mobile as well.

Reviewer 3: Nit: 
```suggestion
.NET Core 6 will allow us to run natively on mobile. Since we are making these changes before .NET 6 is released, I propose we don't include that work as of yet. As long as we handle the native binaries correctly and make sure ML.NET provides descriptive error methods, we should be able to have mobile support as soon as .NET Core 6 releases for everything that currently has a software fallback. Since the native projects we are proposing to keep build for Arm64, they should work on mobile as well.
```
